File: src/claw_mem/merge/conflict_detector.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
ConflictDetector (F3 · v4.7.0)

Detects contradictory or inconsistent semantic memories and resolves them
by keeping the higher-confidence version with optional LLM arbitration.
"""

import re
import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from ..memory_manager import MemoryManager
    from ..llm_provider import LLMProvider
    from ..retrieval.embedding_service import EmbeddingService

from .semantic_merger import _cosine_similarity


# ── data types ─────────────────────────────────────────────────────────

@dataclass
class ConflictReport:
    """A detected conflict between two memory records."""
    conflict_type: str            # "entity", "timeline", "semantic"
    memory_id_a: str
    memory_id_b: str
    content_a: str
    content_b: str
    description: str = ""
    similarity: float = 0.0
    resolved: bool = False
    resolution: Optional["ConflictResolution"] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "conflict_type": self.conflict_type,
            "memory_id_a": self.memory_id_a,
            "memory_id_b": self.memory_id_b,
            "content_a": self.content_a,
            "content_b": self.content_b,
            "description": self.description,
            "similarity": self.similarity,
            "resolved": self.resolved,
            "resolution": self.resolution.to_dict() if self.resolution else None,
        }


@dataclass
class ConflictResolution:
    """Resolution of a detected conflict."""
    action: str                  # "keep_a", "keep_b", "merge", "manual"
    winner_id: str = ""          # Which memory to keep (if keep_*)
    merged_content: str = ""     # Merged text (if action=merge)
    reasoning: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "action": self.action,
            "winner_id": self.winner_id,
            "merged_content": self.merged_content,
            "reasoning": self.reasoning,
        }


# ── LLM prompts ────────────────────────────────────────────────────────

_CONFLICT_CHECK_PROMPT = (
    "Are these two statements contradictory or inconsistent? "
    "Answer YES or NO, then explain briefly.\n\n"
    "Statement A: {a}\nStatement B: {b}"
)

_CONFLICT_RESOLVE_PROMPT = (
    "Choose the correct version from two conflicting memories. "
    "Pick the one that is more specific, more recent, or more authoritative. "
    "Reply with 'A', 'B', or 'MERGE'. If MERGE, provide the merged text.\n\n"
    "Memory A: {a}\nMemory B: {b}\n\nAnswer:"
)

# ── entity extraction helpers ──────────────────────────────────────────

_ENTITY_PATTERNS = [
    re.compile(r'\b([A-Z][a-z]+(?:\s+[A-Z][a-z]+)*)\b'),       # Proper names
    re.compile(r'@(\w+)'),                                       # @mentions
    re.compile(r'(?:city|country|company|person|project)\s*[:=]\s*(\S+)', re.I),  # key:value
]


def _extract_entities(text: str) -> List[str]:
    """Extract candidate entity names from text using simple heuristics."""
    entities: List[str] = []
    for pattern in _ENTITY_PATTERNS:
        for match in pattern.findall(text):
            name = match.strip().lower()
            if len(name) >= 2 and name not in entities:
                entities.append(name)
    return entities


def _extract_attributes(text: str) -> Dict[str, str]:
    """Extract attribute key-value pairs (e.g., 'age: 30', 'location: Beijing')."""
    attr_pattern = re.compile(r'(\w+)\s*[:=]\s*([^,;.\n]+)', re.I)
    attrs: Dict[str, str] = {}
    for match in attr_pattern.findall(text):
        key = match[0].strip().lower()
        val = match[1].strip()
        attrs[key] = val
    return attrs


# ── ConflictDetector ───────────────────────────────────────────────────

class ConflictDetector:
    """Detects and resolves contradictory semantic memories.

    Three detection strategies:
      1. Entity attribute conflicts: same entity, different attribute values
      2. Timeline conflicts: event time order inconsistencies
      3. Semantic conflicts: highly similar (>0.7) but contradictory content
    """

    def __init__(
        self,
        manager: "MemoryManager",
        llm_provider: "LLMProvider",
        embedding_service: "EmbeddingService" = None,
        sim_threshold: float = 0.7,
    ):
        self.manager = manager
        self.llm_provider = llm_provider
        self._embedding_service = embedding_service
        self.sim_threshold = sim_threshold
        self._conflict_history: List[ConflictReport] = []

    @property
    def embedding_service(self) -> "EmbeddingService":
        if self._embedding_service is None:
            from ..retrieval.embedding_service import EmbeddingService
            self._embedding_service = EmbeddingService()
        return self._embedding_service

    # ── detection ──────────────────────────────────────────────────────

    def detect_conflicts(self) -> List[ConflictReport]:
        """Run all three conflict detection strategies.

        Returns a (possibly empty) list of ConflictReport objects.
        """
        storage = self.manager.semantic
        all_memories = storage.get_all()
        active = [m for m in all_memories
                  if m.get("id") and m.get("content")
                  and m.get("metadata", {}).get("deprecated") not in ("true", "True", "1")]

        if len(active) < 2:
            return []

        conflicts: List[ConflictReport] = []

        # Strategy 1: entity attribute conflicts
        conflicts.extend(self._detect_entity_conflicts(active))

        # Strategy 2: timeline conflicts
        conflicts.extend(self._detect_timeline_conflicts(active))

        # Strategy 3: semantic conflicts (high similarity, contradictory)
        conflicts.extend(self._detect_semantic_conflicts(active))

        self._conflict_history.extend(conflicts)
        return conflicts

    def _detect_entity_conflicts(self, active: List[Dict[str, Any]]) -> List[ConflictReport]:
        """Detect same-entity, different-attribute-value conflicts."""
        conflicts: List[ConflictReport] = []

        # Index memories by entities they mention
        entity_index: Dict[str, List[Dict[str, Any]]] = {}
        for mem in active:
            for entity in _extract_entities(mem["content"]):
                entity_index.setdefault(entity, []).append(mem)

        seen_pairs: set = set()
        for entity, mems in entity_index.items():
            if len(mems) < 2:
                continue
            for i in range(len(mems)):
                for j in range(i + 1, len(mems)):
                    mid_a, mid_b = mems[i]["id"], mems[j]["id"]
                    pair_key = tuple(sorted([mid_a, mid_b]))
                    if pair_key in seen_pairs:
                        continue
                    seen_pairs.add(pair_key)

                    attrs_a = _extract_attributes(mems[i]["content"])
                    attrs_b = _extract_attributes(mems[j]["content"])
                    conflicts_found = []
                    for key in set(attrs_a) & set(attrs_b):
                        if attrs_a[key].lower() != attrs_b[key].lower():
                            conflicts_found.append(f"{key}: {attrs_a[key]} vs {attrs_b[key]}")

                    if conflicts_found:
                        conflicts.append(ConflictReport(
                            conflict_type="entity",
                            memory_id_a=mid_a,
                            memory_id_b=mid_b,
                            content_a=mems[i]["content"],
                            content_b=mems[j]["content"],
                            description=f"Entity '{entity}' has conflicting attributes: "
                                        f"{', '.join(conflicts_found)}",
                        ))
        return conflicts

    def _detect_timeline_conflicts(self, active: List[Dict[str, Any]]) -> List[ConflictReport]:
        """Detect timeline inconsistencies (event A happens before B but described differently)."""
        conflicts: List[ConflictReport] = []

        # Find memories with temporal expressions
        time_pattern = re.compile(
            r'(?:in|at|on|during|since|before|after|until|from|by)\s+'
            r'(?:the\s+)?(?:Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|'
            r'Jul(?:y)?|Aug(?:ust)?|Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)'
            r'\s+\d{2,4}|\d{2,4}[-\/]\d{1,2}[-\/]\d{1,2}', re.I
        )

        timed_memories = [m for m in active if time_pattern.search(m["content"])]
        if len(timed_memories) < 2:
            return conflicts

        # Compare pairs with temporal references via LLM
        seen_pairs: set = set()
        for i in range(len(timed_memories)):
            for j in range(i + 1, len(timed_memories)):
                mid_a, mid_b = timed_memories[i]["id"], timed_memories[j]["id"]
                pair_key = tuple(sorted([mid_a, mid_b]))
                if pair_key in seen_pairs:
                    continue
                seen_pairs.add(pair_key)

                prompt = _CONFLICT_CHECK_PROMPT.format(
                    a=timed_memories[i]["content"],
                    b=timed_memories[j]["content"],
                )
                response = self.llm_provider.generate(prompt, max_tokens=64)
                if response.upper().startswith("YES"):
                    conflicts.append(ConflictReport(
                        conflict_type="timeline",
                        memory_id_a=mid_a,
                        memory_id_b=mid_b,
                        content_a=timed_memories[i]["content"],
                        content_b=timed_memories[j]["content"],
                        description=response,
                    ))

        return conflicts

    def _detect_semantic_conflicts(self, active: List[Dict[str, Any]]) -> List[ConflictReport]:
        """Detect highly similar but contradictory memories via embeddings + LLM."""
        conflicts: List[ConflictReport] = []
        if len(active) < 2:
            return conflicts

        texts = [m["content"] for m in active]
        try:
            embeddings = self.embedding_service.encode(texts)
        except Exception:
            return conflicts

        n = len(active)
        seen_pairs: set = set()
        for i in range(n):
            for j in range(i + 1, n):
                mid_a, mid_b = active[i]["id"], active[j]["id"]
                pair_key = tuple(sorted([mid_a, mid_b]))
                if pair_key in seen_pairs:
                    continue
                seen_pairs.add(pair_key)

                sim = _cosine_similarity(embeddings[i], embeddings[j])
                if sim < self.sim_threshold:
                    continue

                # High similarity — check for contradiction via LLM
                prompt = _CONFLICT_CHECK_PROMPT.format(
                    a=active[i]["content"],
                    b=active[j]["content"],
                )
                response = self.llm_provider.generate(prompt, max_tokens=64)
                if response.upper().startswith("YES"):
                    conflicts.append(ConflictReport(
                        conflict_type="semantic",
                        memory_id_a=mid_a,
                        memory_id_b=mid_b,
                        content_a=active[i]["content"],
                        content_b=active[j]["content"],
                        description=response,
                        similarity=sim,
                    ))

        return conflicts

    # ── resolution ─────────────────────────────────────────────────────

    def resolve_conflict(self, conflict: ConflictReport) -> ConflictResolution:
        """Resolve a single conflict via LLM arbitration.

        Strategy: keep the higher-confidence version, or merge if equally valid.
        """
        prompt = _CONFLICT_RESOLVE_PROMPT.format(
            a=conflict.content_a, b=conflict.content_b
        )
        answer = self.llm_provider.generate(prompt, max_tokens=128)

        if answer.upper().startswith("A"):
            resolution = ConflictResolution(
                action="keep_a",
                winner_id=conflict.memory_id_a,
                reasoning=answer,
            )
        elif answer.upper().startswith("B"):
            resolution = ConflictResolution(
                action="keep_b",
                winner_id=conflict.memory_id_b,
                reasoning=answer,
            )
        else:
            # Default to merge — try to combine
            merged = self.llm_provider.generate(
                f"Combine these two statements into one accurate statement. "
                f"Resolve any contradictions by keeping the more specific version.\n\n"
                f"A: {conflict.content_a}\nB: {conflict.content_b}\n\nCombined:",
                max_tokens=256,
            )
            resolution = ConflictResolution(
                action="merge",
                merged_content=merged or conflict.content_a,
                reasoning=answer,
            )

        conflict.resolved = True
        conflict.resolution = resolution
        return resolution

    # ── full cycle ─────────────────────────────────────────────────────

    def run_cycle(self) -> Dict[str, Any]:
        """Run a full conflict detection and resolution cycle.

        Returns stats dict.
        """
        t0 = time.monotonic()
        conflicts = self.detect_conflicts()

        resolved_count = 0
        for conflict in conflicts:
            try:
                self.resolve_conflict(conflict)
                resolved_count += 1
            except Exception:
                pass

        duration = round((time.monotonic() - t0) * 1000, 1)

        return {
            "conflicts_detected": len(conflicts),
            "conflicts_resolved": resolved_count,
            "by_type": {
                "entity": sum(1 for c in conflicts if c.conflict_type == "entity"),
                "timeline": sum(1 for c in conflicts if c.conflict_type == "timeline"),
                "semantic": sum(1 for c in conflicts if c.conflict_type == "semantic"),
            },
            "duration_ms": duration,
        }

    def get_history(self) -> List[ConflictReport]:
        """Return the full conflict detection history."""
        return list(self._conflict_history)

    def clear_history(self) -> None:
        """Clear the conflict history."""
        self._conflict_history.clear()

    def __repr__(self) -> str:
        return f"ConflictDetector(threshold={self.sim_threshold})"

File: src/merge/conflict_detector.ts
// Copyright 2026 Peter Cheng
// Licensed under Apache-2.0

/**
 * ConflictDetector (F3 - v4.7.0)
 *
 * Detects contradictory or inconsistent semantic memories and resolves them
 * by keeping the higher-confidence version with optional LLM arbitration.
 */

import { cosineSimilarity } from "./semantic_merger.js";

// ── External type stubs ────────────────────────────────────────────────

interface StorageBackend {
  getAll(): Record<string, unknown>[];
}

interface LLMProvider {
  generate(prompt: string, options?: { maxTokens?: number; system?: string }): string;
}

interface EmbeddingService {
  encode(texts: string[]): number[][];
}

interface MemoryManager {
  semantic: StorageBackend;
}

// ── Data types ─────────────────────────────────────────────────────────

export interface ConflictReport {
  conflictType: string; // "entity" | "timeline" | "semantic"
  memoryIdA: string;
  memoryIdB: string;
  contentA: string;
  contentB: string;
  description: string;
  similarity: number;
  resolved: boolean;
  resolution: ConflictResolution | null;
}

export function conflictReportToDict(r: ConflictReport): Record<string, unknown> {
  return {
    conflict_type: r.conflictType,
    memory_id_a: r.memoryIdA,
    memory_id_b: r.memoryIdB,
    content_a: r.contentA,
    content_b: r.contentB,
    description: r.description,
    similarity: r.similarity,
    resolved: r.resolved,
    resolution: r.resolution ? conflictResolutionToDict(r.resolution) : null,
  };
}

export interface ConflictResolution {
  action: string; // "keep_a" | "keep_b" | "merge" | "manual"
  winnerId: string;
  mergedContent: string;
  reasoning: string;
}

export function conflictResolutionToDict(r: ConflictResolution): Record<string, unknown> {
  return {
    action: r.action,
    winner_id: r.winnerId,
    merged_content: r.mergedContent,
    reasoning: r.reasoning,
  };
}

// ── LLM prompts ────────────────────────────────────────────────────────

const CONFLICT_CHECK_PROMPT =
  "Are these two statements contradictory or inconsistent? " +
  "Answer YES or NO, then explain briefly.\n\n" +
  "Statement A: {a}\nStatement B: {b}";

const CONFLICT_RESOLVE_PROMPT =
  "Choose the correct version from two conflicting memories. " +
  "Pick the one that is more specific, more recent, or more authoritative. " +
  "Reply with 'A', 'B', or 'MERGE'. If MERGE, provide the merged text.\n\n" +
  "Memory A: {a}\nMemory B: {b}\n\nAnswer:";

// ── entity extraction helpers ──────────────────────────────────────────

/** Extract candidate entity names from text using simple heuristics. */
function extractEntities(text: string): string[] {
  const entities: string[] = [];
  const patterns: RegExp[] = [
    /\b([A-Z][a-z]+(?:\s+[A-Z][a-z]+)*)\b/g,
    /@(\w+)/g,
    /(?:city|country|company|person|project)\s*[:=]\s*(\S+)/gi,
  ];
  for (const pattern of patterns) {
    let match: RegExpExecArray | null;
    while ((match = pattern.exec(text)) !== null) {
      const name = match[1].trim().toLowerCase();
      if (name.length >= 2 && !entities.includes(name)) {
        entities.push(name);
      }
    }
  }
  return entities;
}

/** Extract attribute key-value pairs (e.g., "age: 30", "location: Beijing"). */
function extractAttributes(text: string): Record<string, string> {
  const attrPattern = /(\w+)\s*[:=]\s*([^,;.\n]+)/gi;
  const attrs: Record<string, string> = {};
  let match: RegExpExecArray | null;
  while ((match = attrPattern.exec(text)) !== null) {
    const key = match[1].trim().toLowerCase();
    const val = match[2].trim();
    attrs[key] = val;
  }
  return attrs;
}

// ── ConflictDetector ───────────────────────────────────────────────────

export class ConflictDetector {
  private _conflictHistory: ConflictReport[] = [];

  constructor(
    private _manager: MemoryManager,
    private _llmProvider: LLMProvider,
    private _embeddingService?: EmbeddingService,
    public simThreshold: number = 0.7,
  ) {}

  private get embeddingService(): EmbeddingService {
    if (!this._embeddingService) {
      throw new Error("EmbeddingService not configured");
    }
    return this._embeddingService;
  }

  // ── detection ────────────────────────────────────────────────────────

  /** Run all three conflict detection strategies. Returns a (possibly empty) list. */
  detectConflicts(): ConflictReport[] {
    const storage = this._manager.semantic;
    const allMemories = storage.getAll();
    const active = allMemories.filter(
      (m) =>
        m.id &&
        m.content &&
        (m.metadata as Record<string, string>)?.deprecated !== "true" &&
        (m.metadata as Record<string, string>)?.deprecated !== "True" &&
        (m.metadata as Record<string, string>)?.deprecated !== "1",
    ) as Array<Record<string, string>>;

    if (active.length < 2) return [];

    const conflicts: ConflictReport[] = [];

    // Strategy 1: entity attribute conflicts
    conflicts.push(...this._detectEntityConflicts(active));

    // Strategy 2: timeline conflicts
    conflicts.push(...this._detectTimelineConflicts(active));

    // Strategy 3: semantic conflicts (high similarity, contradictory)
    conflicts.push(...this._detectSemanticConflicts(active));

    this._conflictHistory.push(...conflicts);
    return conflicts;
  }

  /** Detect same-entity, different-attribute-value conflicts. */
  private _detectEntityConflicts(
    active: Array<Record<string, string>>,
  ): ConflictReport[] {
    const conflicts: ConflictReport[] = [];

    // Index memories by entities they mention
    const entityIndex: Record<string, Array<Record<string, string>>> = {};
    for (const mem of active) {
      for (const entity of extractEntities(mem.content)) {
        if (!entityIndex[entity]) entityIndex[entity] = [];
        entityIndex[entity].push(mem);
      }
    }

    const seenPairs = new Set<string>();
    for (const [entity, mems] of Object.entries(entityIndex)) {
      if (mems.length < 2) continue;
      for (let i = 0; i < mems.length; i++) {
        for (let j = i + 1; j < mems.length; j++) {
          const midA = mems[i].id;
          const midB = mems[j].id;
          const pairKey = [midA, midB].sort().join("|");
          if (seenPairs.has(pairKey)) continue;
          seenPairs.add(pairKey);

          const attrsA = extractAttributes(mems[i].content);
          const attrsB = extractAttributes(mems[j].content);
          const conflictsFound: string[] = [];
          for (const key of Object.keys(attrsA)) {
            if (key in attrsB && attrsA[key].toLowerCase() !== attrsB[key].toLowerCase()) {
              conflictsFound.push(`${key}: ${attrsA[key]} vs ${attrsB[key]}`);
            }
          }

          if (conflictsFound.length > 0) {
            conflicts.push({
              conflictType: "entity",
              memoryIdA: midA,
              memoryIdB: midB,
              contentA: mems[i].content,
              contentB: mems[j].content,
              description: `Entity '${entity}' has conflicting attributes: ${conflictsFound.join(", ")}`,
              similarity: 0,
              resolved: false,
              resolution: null,
            });
          }
        }
      }
    }
    return conflicts;
  }

  /** Detect timeline inconsistencies (event order described differently). */
  private _detectTimelineConflicts(
    active: Array<Record<string, string>>,
  ): ConflictReport[] {
    const conflicts: ConflictReport[] = [];

    // Find memories with temporal expressions
    const timePattern = new RegExp(
      "(?:in|at|on|during|since|before|after|until|from|by)\\s+" +
        "(?:the\\s+)?(?:Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|" +
        "Jul(?:y)?|Aug(?:ust)?|Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)" +
        "\\s+\\d{2,4}|\\d{2,4}[-/]\\d{1,2}[-/]\\d{1,2}",
      "i",
    );

    const timedMemories = active.filter((m) => timePattern.test(m.content));
    if (timedMemories.length < 2) return conflicts;

    // Compare pairs with temporal references via LLM
    const seenPairs = new Set<string>();
    for (let i = 0; i < timedMemories.length; i++) {
      for (let j = i + 1; j < timedMemories.length; j++) {
        const midA = timedMemories[i].id;
        const midB = timedMemories[j].id;
        const pairKey = [midA, midB].sort().join("|");
        if (seenPairs.has(pairKey)) continue;
        seenPairs.add(pairKey);

        const prompt = CONFLICT_CHECK_PROMPT.replace(
          "{a}",
          timedMemories[i].content,
        ).replace("{b}", timedMemories[j].content);
        const response = this._llmProvider.generate(prompt, { maxTokens: 64 });
        if (response.toUpperCase().startsWith("YES")) {
          conflicts.push({
            conflictType: "timeline",
            memoryIdA: midA,
            memoryIdB: midB,
            contentA: timedMemories[i].content,
            contentB: timedMemories[j].content,
            description: response,
            similarity: 0,
            resolved: false,
            resolution: null,
          });
        }
      }
    }

    return conflicts;
  }

  /** Detect highly similar but contradictory memories via embeddings + LLM. */
  private _detectSemanticConflicts(
    active: Array<Record<string, string>>,
  ): ConflictReport[] {
    const conflicts: ConflictReport[] = [];
    if (active.length < 2) return conflicts;

    const texts = active.map((m) => m.content);
    let embeddings: number[][];
    try {
      embeddings = this.embeddingService.encode(texts);
    } catch {
      return conflicts;
    }

    const n = active.length;
    const seenPairs = new Set<string>();
    for (let i = 0; i < n; i++) {
      for (let j = i + 1; j < n; j++) {
        const midA = active[i].id;
        const midB = active[j].id;
        const pairKey = [midA, midB].sort().join("|");
        if (seenPairs.has(pairKey)) continue;
        seenPairs.add(pairKey);

        const sim = cosineSimilarity(embeddings[i], embeddings[j]);
        if (sim < this.simThreshold) continue;

        // High similarity -- check for contradiction via LLM
        const prompt = CONFLICT_CHECK_PROMPT.replace("{a}", active[i].content).replace(
          "{b}",
          active[j].content,
        );
        const response = this._llmProvider.generate(prompt, { maxTokens: 64 });
        if (response.toUpperCase().startsWith("YES")) {
          conflicts.push({
            conflictType: "semantic",
            memoryIdA: midA,
            memoryIdB: midB,
            contentA: active[i].content,
            contentB: active[j].content,
            description: response,
            similarity: sim,
            resolved: false,
            resolution: null,
          });
        }
      }
    }

    return conflicts;
  }

  // ── resolution ───────────────────────────────────────────────────────

  /** Resolve a single conflict via LLM arbitration. */
  resolveConflict(conflict: ConflictReport): ConflictResolution {
    const prompt = CONFLICT_RESOLVE_PROMPT.replace("{a}", conflict.contentA).replace(
      "{b}",
      conflict.contentB,
    );
    const answer = this._llmProvider.generate(prompt, { maxTokens: 128 });

    let resolution: ConflictResolution;
    if (answer.toUpperCase().startsWith("A")) {
      resolution = {
        action: "keep_a",
        winnerId: conflict.memoryIdA,
        mergedContent: "",
        reasoning: answer,
      };
    } else if (answer.toUpperCase().startsWith("B")) {
      resolution = {
        action: "keep_b",
        winnerId: conflict.memoryIdB,
        mergedContent: "",
        reasoning: answer,
      };
    } else {
      // Default to merge -- try to combine
      const merged = this._llmProvider.generate(
        `Combine these two statements into one accurate statement. ` +
          `Resolve any contradictions by keeping the more specific version.\n\n` +
          `A: ${conflict.contentA}\nB: ${conflict.contentB}\n\nCombined:`,
        { maxTokens: 256 },
      );
      resolution = {
        action: "merge",
        winnerId: "",
        mergedContent: merged || conflict.contentA,
        reasoning: answer,
      };
    }

    conflict.resolved = true;
    conflict.resolution = resolution;
    return resolution;
  }

  // ── full cycle ───────────────────────────────────────────────────────

  /** Run a full conflict detection and resolution cycle. Returns stats dict. */
  runCycle(): Record<string, unknown> {
    const t0 = Date.now();
    const conflicts = this.detectConflicts();

    let resolvedCount = 0;
    for (const conflict of conflicts) {
      try {
        this.resolveConflict(conflict);
        resolvedCount++;
      } catch {
        // skip failed resolutions
      }
    }

    const duration = Math.round(Date.now() - t0);

    return {
      conflictsDetected: conflicts.length,
      conflictsResolved: resolvedCount,
      byType: {
        entity: conflicts.filter((c) => c.conflictType === "entity").length,
        timeline: conflicts.filter((c) => c.conflictType === "timeline").length,
        semantic: conflicts.filter((c) => c.conflictType === "semantic").length,
      },
      durationMs: duration,
    };
  }

  /** Return the full conflict detection history. */
  getHistory(): ConflictReport[] {
    return [...this._conflictHistory];
  }

  /** Clear the conflict history. */
  clearHistory(): void {
    this._conflictHistory = [];
  }

  toString(): string {
    return `ConflictDetector(threshold=${this.simThreshold})`;
  }
}
